code/game/world_topic_commands.dm

```diff
@@ -0,0 +1,363 @@
+var/global/list/decl/topic_command/topic_commands = list()
+
+/decl/topic_command
+	var/name = "topiccommand"
+	var/has_params = FALSE
+	var/base_type = /decl/topic_command
+
+/// Initialises the assoc list of topic commands by key.
+/hook/startup/proc/setup_api()
+	for (var/decl/topic_command/TC in decls_repository.get_decls_of_subtype(/decl/topic_command))
+		if(TC.base_type == type)
+			continue
+		global.topic_commands[TC.name] = TC
+	return TRUE
+
+/// Returns a truthy value if we CANNOT use this command.
+/decl/topic_command/proc/can_use(var/T, var/addr, var/master, var/key)
+	if(base_type == type) // this is an abstract type
+		return FALSE
+	if (has_params)
+		if (copytext(T, 1, length(name) + 1) != name)
+			return FALSE
+	else if (T != name)
+		return FALSE
+	return TRUE
+
+/decl/topic_command/proc/use(var/list/params)
+	return FALSE
+
+/decl/topic_command/proc/try_use(var/T, var/addr, var/master, var/key)
+	if (!can_use(T, addr, master, key))
+		return FALSE
+	var/list/params
+	if(has_params)
+		params = params2list(T)
+	return use(params)
+
+/decl/topic_command/secure
+	base_type = /decl/topic_command/secure
+
+/decl/topic_command/secure/try_use(var/T, var/addr, var/master, var/key)
+	if (!can_use(T, addr, master, key))
+		return FALSE
+	var/list/params = params2list(T)
+	if (!config.comms_password)
+		set_throttle(addr, 10 SECONDS, "Comms Not Enabled")
+		return "Not Enabled"
+	if (params["key"] != config.comms_password)
+		set_throttle(addr, 30 SECONDS, "Bad Comms Key")
+		return "Bad Key"
+	return use(params)
+
+/* * * * * * * *
+* Public Topic Calls
+* The following topic calls are available without a comms secret.
+* * * * * * * */
+
+/decl/topic_command/ping
+	name = "ping"
+
+/decl/topic_command/ping/use()
+	var/x = 1
+	for (var/client/C)
+		x++
+	return x
+
+/decl/topic_command/players
+	name = "players"
+
+/decl/topic_command/players/use()
+	var/n = 0
+	for(var/mob/M in global.player_list)
```

Comment:
ayy lmao legacy code
the entire thing can be replaced with `global.clients || 0`